### ISSUE TITLE
Got type INTEGER for field `hd_buy_potential` at position 2, but expected VARCHAR

### ISSUE THREAD

**Issue by Guhaiyan** [facebookincubator/velox#3084]
I encountered error below when running TPC-DC q73 with duckdb parquet reader and HiveConnector in Debug Version(checks producing the error is only done in Debug,  different error will show in Release)。

`failed: child->type()->kindEquals(type->childAt(i)) Got type INTEGER for field `hd_buy_potential` at position 2, but expected VARCHAR.Split [file hdfs://localhost:9000/user/hive/warehouse/tpcds.db/household_demographics/20221020_031627_00008_bq3ke_824562d0-e76c-4940-89cb-fcd89eabfb9f 0 - 12634] Task 20221103_080422_00019_cx9ut.10.0.0`

The simplified case is
`select  hd_demo_sk from household_demographics where hd_buy_potential = 'Unknown' and hd_vehicle_count + rand() > 0;`

If `WHERE `predicate matches the pattern that filters does not project out and remainingFilter is not empty, error occurs。In the example sql above, hd_buy_potential is filter and it is not in `SELECT `list,  while hd_vehicle_count is remainingFilter.

My analysis for this problem is illustrated below.

Columns' position in readerOutputType_ is supposed to like this, projected-out columns put first continuously and then no-projected-out columns,  so that  VELOX_DCHECK in RowVector constructor can pass。There should not have any no-projected-out column between  projected-out columns, otherwise you will get this error in Debug.

The field sequence in ScanSpec now is `[output columns, filters(not duplicated),  remainingFilter]`, and extra filters columns( not duplicated with output columns) are not projected out。If remainingFilter , which needs to be projected internally for further evalulation, is not empty, there will be gap( null child) between none-empty children in RowVector, the VELOX_DCHECK fails。

One possible solution to this problem is switching filters and remainingFilter 's position in ScanSpec, making field sequence in ScanSpec to be `[output columns, filters(not duplicated),  remainingFilter]`.

**Comment by mbasmanova** [comment]
@Guhaiyan Thank you for reporting this issue. Can you write a test case that reproduces this problem?

CC: @yingsu00 @oerling @Yuhta

**Comment by mbasmanova** [comment]
@Guhaiyan Would you attach full stack trace?

**Comment by Guhaiyan** [comment]
> @Guhaiyan Thank you for reporting this issue. Can you write a test case that reproduces this problem?
> 
> CC: @yingsu00 @oerling @Yuhta

The query to reproduce this problem is quite simple。
steps:
```
create table tt(c1 int, c2 varchar, c3 date);
insert into tt values(1,'a', date'2022-11-03');
insert into tt values(2,'b', date'2022-11-04');
select c3 from tt where c2 = 'a' and c1 + 1 > 0;
```
I am not sure whether it is easy to reproduce it using PlanBuilder,  because PlanBuilder.tableScan() requires subfieldFileters be one of outputType, which is more like select c3, c2 from tt where c2, even though you can project c3 only by PlanBuilder.tableScan().project()。

**Comment by oerling** [comment]
How does this work with the native Parquet reader? Is there a particular reason you are running the DuckDB one? If so, we should fix the native one first. We are likely to stop maintaining the DuckDB reader.

**Comment by Yuhta** [comment]
@Guhaiyan Do you mean changing the `ScanSpec`s to the order `[output columns, remainingFilter, filters(not duplicated)]`?  Can you summarize this logic in the description of PR and put some test?

**Comment by Guhaiyan** [comment]
> How does this work with the native Parquet reader? 
I run the same test case with native one, fortunately it works.  
In this test case, the main difference between duckdb and native is that native does no care about spec->projectOut() flag but duckdb does.  Duckdb choose projected-out spec to build VeloxVector using duckdb::toVeloxVector.

```
 } else if (spec->projectOut()) {
        auto index = rowType_->getChildIdx(spec->fieldName());
        columns[spec->channel()] = duckdb::toVeloxVector(
            output.size(),
            output.data[index],
            rowType_->childAt(index),
            &pool_);
```

Is there a particular reason you are running the DuckDB one? If so, we should fix the native one first. We are likely to stop maintaining the DuckDB reader.
The reason to  use DuckDB  reader is quite simple，there is no native parquet reader yet when we first run TPC-DS test. And if duckdb reader is no longer maintained as well as native parquet reader now is ready and stable, it seems switch to native one from duckdb is a must choice.
